### ISSUE TITLE
safari: infinite loop when hot reloading

### ISSUE THREAD

**Issue by ydnar** [nuxt/nuxt.js#3828]
### Version

[v1.4.2](https://github.com/nuxt.js/releases/tag/v1.4.2) (or latest 2.0.0 pre code)

### Steps to reproduce

1. Launch dev server via `yarn dev` (`HOST=0.0.0.0 nuxt`)
2. Load http://localhost:3000/
3. Edit code, any code. Watch the nuxt dev server recompile everything.
4. Reload
5. HMR fails, triggering rapid reloading of the entire page, forever, until Safari kills the tab for excessive CPU usage.
6. The only way to test new code is to close the tab and open a new tab to localhost (and re-open dev console).

### What is expected ?

Hot reload to work without forcing me to close the tab and open a new tab every time I change a line of code.

### What is actually happening?

Browser crushing infinite reload.

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c7666">#c7666</a>)</em></sub></div>

**Comment by manniL** [comment]
@ydnar Have you tried to remove all page information about localhost in safari? I had similar issues when a service worker stuck.

**Comment by aldarund** [comment]
I have same\similar issue on chrome all time. I have use of nuxt/pwa. And reloading just  hangup on some of hot reload and after it page dont load at all no matter if u close tab, open new tab or anything it will just load forever. Only anonymous window helps. Pretty terrible dev experience

**Comment by manniL** [comment]
@aldarund See https://github.com/nuxt-community/pwa-module/pull/60#issuecomment-407008485. This should be an issue of `pwa-module`, not nuxt-core itself.

@ydnar Please ping me if you **do not** use workbox/pwa-module

**Comment by ydnar** [comment]
@manniL we’re not using `pwa-module`, just vanilla Nuxt (`dev` branch).

**Comment by Aric-sky** [comment]
> I have same\similar issue on chrome all time. I have use of nuxt/pwa. And reloading just hangup on some of hot reload and after it page dont load at all no matter if u close tab, open new tab or anything it will just load forever. Only anonymous window helps. Pretty terrible dev experience

Have you solved this problem?

**Comment by Norbz** [comment]
We have the same issue here, some things I tried without any success : 
- Removing the .nuxt folder
- Completely killing the server and restarting
- clearing all cookie, localStorage and stuff...
- reinstalling all node_modules

I am working on MacOS, happens to all browsers, but not all the time. I can sometime works for a few hours before it happens.
I have PWA module installed, but if i record correctly, it happened before I added it to the project (not sure though)

Environment :
- node v8.11.3
- npm v5.5.1
- yarn v1.9.2

Nuxt : v1.4.2
PWA module : v2.5.0

**Comment by galvez** [comment]
I have just experienced the same issue - I am investigating it.

**Comment by manniL** [comment]
Not stale :(

**Comment by manniL** [comment]
We still need a way to reliably reproduce this I think :|

**Comment by yoyo837** [comment]
Same issue here. (safari only) 
I am using safari for the first time today to debug nuxt.js, I found it so late.
![default](https://user-images.githubusercontent.com/6134547/49274217-9c64ae00-f4b2-11e8-8f46-a64e7b5d8024.gif)

**Comment by bovas85** [comment]
is there a fix for this? Safari is breaking on prod for me

**Comment by manniL** [comment]
This shouldn't happen on prod 😱 

@bovas85 can you reproduce it somehow?

**Comment by smyth64** [comment]
I have the same issue, only on Safari 😅

This issue is very important, if it even breaks prod o_O

**Comment by papakay** [comment]
Please i'm having the same issue on Safari 12.0

**Comment by felixdenoix** [comment]
Same here, fyi it also happends for me on a nextjs project !

**Comment by sabithpocker** [comment]
Similar issue with nuxt, Safari shows following errors:

> [Info] You are running Vue in development mode. (commons.app.js, line 17101)
> Make sure to turn on production mode when deploying for production.
> See more tips at https://vuejs.org/guide/deployment.html
> [Log] [HMR] connected (vendors.app.js, line 1830)
> [Log] [HMR] Checking for updates on the server... (vendors.app.js, line 2118)
> [Error] Failed to load resource: the server responded with a status of 404 (Not Found) (3e9ec34fb6b7af1e692e.hot-update.json, line 0)
> [Warning] [HMR] Cannot find update (Full reload needed) (vendors.app.js, line 2128)
> [Warning] [HMR] (Probably because of restarting the server) (vendors.app.js, line 2129)
> [Warning] [HMR] Reloading page (vendors.app.js, line 2219)

<img width="1613" alt="screenshot 2018-12-29 at 10 02 08 pm" src="https://user-images.githubusercontent.com/1050961/50540319-a4009b00-0bb5-11e9-9bd1-72415c754a08.png">

**Comment by valera-rozuvan** [comment]
Started testing Safari, `"nuxt": "2.3.4"`, same problem. This happens when I restart the app, while it is already open in the browser.

**Comment by schrockwell** [comment]
What is especially frustrating about this problem is that a manual refresh does NOT stop the refresh loop, even with clearing the cache (Cmd-Option-R). The window/tab must be completely closed to stop the loop.

I am also experiencing this on mobile Safari on iOS 12.1.2.

**Comment by danchristian** [comment]
I get this too, same as above, only killing the tab solves it  😕

**Comment by liepaja** [comment]
I started getting this when adding `require('dotenv').config()` to the `nuxt.conf.js `. Does this ring a bell for anyone?

**Comment by manniL** [comment]
Hey :wave: 

Is this still reproducible on v2.4.0?

**Comment by schrockwell** [comment]
Confirmed, the issue still occurs on Nuxt 2.4.0 and Safari 12.0. :(

**Comment by pi0** [comment]
I can confirm @yoyo837:

> Need to have a build error, change file, make a mistake, save file, as you can see that now..

**Comment by pi0** [comment]
The important key for reproduction is:

- Make an error that opens overlay
- Make another error that does not cause reload but overlay
- Fix it

**Comment by pi0** [comment]
#4895 should resolve this problem.

**Comment by lbogdan** [comment]
@pi0 But that does not explain what's happening with the nuxt sandbox, where there's no updates.

@everybody I'm willing to look into this, if you can provide me with a remote connection (TeamViewer or similar) to a machine where this can be easily reproduced.

**Comment by schrockwell** [comment]
@lbogdan I was able to make a minimal Nuxt project that recreates the issue. It includes steps to reproduce (intermittently, in this particular case): https://github.com/schrockwell/nuxt-issue-3828

In summary:

1. `npm run dev` and open the index page in Safari 12
2. Uncomment `// syntax error;` in this file, save it, and observe the expected compilation error
3. Re-comment `// syntax error;` to successfully compile, then stare into the HMR abyss!
4. If you get "Module build failed" at this point, repeat steps 2 and 3 and then refresh the page and try again. In this example repo, it's intermittent to get it to occur.

From what I can tell, it comes down to forcing HMR to do a full reload, but somehow Safari caches and displays the old version of the page, which wants a full HMR reload, and so on. If you get stuck in the reload loop and change the page HTML content, you'll see that it does not update accordingly - you will get either the old version of the page, or maybe quickly flashing between the old and new content, depending.

**Comment by manniL** [comment]
Could it be related to https://github.com/webpack/watchpack/issues/25 ? 🤔

**Comment by manniL** [comment]
Fun fact: This is a bug in safari itself.

https://github.com/vuejs/vue-cli/issues/1132

**Comment by rodrigopedra** [comment]
There is a PR in the vuejs repo addressing this issue:

https://github.com/vuejs/vue-cli/pull/2403

Also, I have this problem in Firefox when I set a preference to block cookies from third-party trackers. If I change to any other values the reload loops no longer occurs

To set this option in Firefox, open Firefox and navigate to:

about:preferences#privacy

Then change Content blocking to custom, check the cookies checkbox and select "third-party trackers" from the combo box next to it.

Maybe in Safari it is related to the tracker prevention feature it has built-in. I don't own a Mac, so I cannot try to disable this feature to test if it solves it.

If this is the cause, maybe we have to change how hot-reloading keeps track of it in the browser.

**Comment by manniL** [comment]
@rodrigopedra So, the thing is we used `hash`es for development mode file names too (in 1.x) but they were causing huge memory leaks. And releasing a "fix" that creates another known problem is not the goal.

**Comment by rodrigopedra** [comment]
Hi @manniL thanks for the response. I wasn't aware of these changes made from the 1.x, although I worked with Nuxt before in a previous project, I was hired just to refactor some components and didn't dig into the inner workings as I wasn't responsible for that.

Now I am in a project where we are using Nuxt 2.x and are building the aoo from scratch, so I got the chance to have a deeper look into it.

I agree that revert a change that solved a memory problem is not something to be pursued, but I don't know if using hashes to avoid caching is the only way to solve this problem. From what I could check the problem occurs when using latest Safari or Firefox with tracking blocked. Maybe there are other config we could hack to have these browsers behave better when this feature is on.

Are there caching related headers we could send with the HMR updates up so these browsers doesn't see this feature as a kind of tracking?

I don't have enough knowledge to tell why this issue happens. But I guess the evidence on blocking trackers is a good clue on how to solve it.

If I can help in any ways please let me know.

**Comment by manniL** [comment]
@rodrigopedra I know @pi0 tried to send all possible combinations of "no cache" headers and Safari consecutively ignored them 🙈 

I think a report to the browser(s) bug tracker is the best we can do for now. It should also apply to more projects than just Nuxt.js but I wasn't able to find similar bug reports besides the one linked above from `vue-cli` (where Evan stated it's a Safari bug as well).

**Workarounds:**

* The Rate Limiting PR (#4895) will "solve" the problem
* You can change the `build.filenames` attribute on your own (example over here: https://github.com/schrockwell/nuxt-issue-3828/pull/1)

**Comment by chinafzy** [comment]
Same problem.
node v8.12.0
Nuxt.js v2.6.2
Safari  v12.0.3

**Comment by manniL** [comment]
Closing as workaround is known and we can't do much more here :| See https://github.com/nuxt/nuxt.js/issues/3828#issuecomment-464675623

**Comment by federicovezzoli** [comment]
Hi guys, sorry but the _build.filenames_ fix breaks everything.
And I really don't understand how to apply the rate limiting PR, tried the serverMiddleware, but breaks everything.

Anyone could shed some light?

thanks

**Comment by aldarund** [comment]
@manniL filenames workaround could lead to memory leaks. Didnt really tested, but webpack warns about it
```  WARN  Notice: Please do not use hash in dev mode to prevent memory leak ```

**Comment by aldarund** [comment]
poor safari users end up with either infinite loop or memory leak

**Comment by smyth64** [comment]
😢

Dmitry Molotkov <notifications@github.com> schrieb am Sa. 18. Mai 2019 um
17:31:

**Comment by aldarund** [comment]
fyi - i have did some testing and with build.standalone: true there seem to be no memory leak with hash

**Comment by ladislavsulc** [comment]
I had the same problem now with the latest version in **Opera** (in anonymous mode it works) and this helped: https://user-images.githubusercontent.com/13877593/43817835-66dc5e94-9a90-11e8-84a1-6e42e06e2015.png.

**Comment by tbredin** [comment]
I have this issue in latest Chrome. 

A clue: it doesn't happen if I open it in an incognito window ... very new to nuxt so having a very tough time debugging this

```
nuxt: "^2.4.0"
node: v10.15.3
Chrome: 74.0.3729.169
```

**Comment by valera-rozuvan** [comment]
@manniL Maybe it's time to re-open this issue?

**Comment by ChristianKienle** [comment]
I also just ran into the same issue (I think). I can reproduce it with a minimal project. It basically just is 

```
yarn create nuxt-app
```

and it happens as soon I create the first page.

**Comment by agcty** [comment]
I also experience this issue but only with Safari

**Comment by Ic3m4n34** [comment]
> I also experience this issue but only with Safari

Me too!

**Comment by valera-rozuvan** [comment]
This issues will soon reach 1 year old mark. It still annoys devs :) @manniL maybe time to re-open and re-investigate?

**Comment by valera-rozuvan** [comment]
@pi0 , @clarkdo : do you guys experience similar issue on your local dev machine?

**Comment by clarkdo** [comment]
@valera-rozuvan As @manniL 's comment above https://github.com/nuxt/nuxt.js/issues/3828#issuecomment-464675623, it's actually a safari bug, as a workaround you may try to use @pi0 's solution.

Or change filename in dev mode, like:

```js
// nuxt.config.js
export default {
  build: {
    filenames: {
      app: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js',
      chunk: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js'
    }
  }
}


```

**Comment by FeatureSpitter** [comment]
This solution above worked for me, and I was having this same problem in Chrome 75.0.3770.100 (64 bits), when using NGINX to expose my `npm run dev` as HTTPS (more details here: https://stackoverflow.com/questions/56966137/how-to-run-nuxt-npm-run-dev-with-https-in-localhost).

Thanks @clarkdo !

**Comment by Leskd** [comment]
@clarkdo It's not worked for me. I was having this same problem in Chrome 75.0.3770.100.
```
nuxt: "^2.8.1"
node: v10.15.3
Chrome: 74.0.3729.169
```

**Comment by clarkdo** [comment]
Could you provide a reproducible repo?

**Comment by Leskd** [comment]
@clarkdo I will provide it. wait a moment please

**Comment by Leskd** [comment]
@clarkdo It's weird when I am prepaing a reproducible repo. I test again, It's normal now. Only I deleted is the .nuxt folder.  then I deleted
```
filenames: {
      app: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js',
      chunk: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js'
    }
```
too.
It's work.

**Comment by preom** [comment]
FWIW I am in a similar boat to https://github.com/nuxt/nuxt.js/issues/3828#issuecomment-498518622. I am also getting this issue in chrome but not in incognito mode. I've already tried isolating the problem by process of elimination but the problem seems to come and go. I don't really want to waste anymore time on this since there is a solution (incognito mode) and this doesn't seem to be affecting other people.

**Comment by updatesvc** [comment]
This might help someone...
How i solved it:

- Go to devtools
- Go to Application Tab
- On the left panel click clear storage
- Then on right-click, clear site data

**Comment by istvanvasil** [comment]
Had the same issue in Chrome until I updated it. Must have been linked to that specific version. Otherwise `filenames: {
      app: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js',
      chunk: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js'
    }` solves this problem

**Comment by AJB99** [comment]
Same issue here.

Is there any plan to resolve this?
Has this been accepted as a bug?

Nuxt is not an option if this keeps happening.

**Comment by JakubKoralewski** [comment]
Happens for me on Edge ![Edge](https://i.imgur.com/luPwTzM.gif), Firefox ![Firefox](https://i.imgur.com/xExRtSr.gif), Chrome ![Chrome](https://i.imgur.com/SflrrBP.gif), Safari ![Safari](https://i.imgur.com/mlpwBgW.gif)

nuxt 2.11.0

**Comment by kkostov** [comment]
A "safari bug" was mentioned above, could anyone point to it on the webkit bug tracker? Or was there a radar report to Apple? I tried to look it up but frankly I don't understand webpack enough to know what to look for...

**Comment by hfnukal** [comment]
Same issue. I have to press [CMD]+[e] (clear cache) [CMD]+[r] (reload)

**Comment by airterjun** [comment]
Hi guys, is someone finds the solution, I have the same issue here, 
It's happening when I added middleware into my nuxt.config.js

nuxt.config.js
`
...
 router: {
     middleware: 'checker'
 }
...
`
checker.js
`
export default function (context) {
   console.log("param", context.params)
}
`

my page route 
![image](https://user-images.githubusercontent.com/24951/78605378-fab35500-788d-11ea-8aa3-bc6dc65cb353.png)

![image](https://user-images.githubusercontent.com/24951/78605441-1a4a7d80-788e-11ea-84fe-76c3a8edade7.png)

Thanks

**Comment by DengSihan** [comment]
Same issue

Only `nuxt.config.js` but issue still happened in Chrome and Firefox
```js
module.exports = {
    srcDir: __dirname,
    watchers: {
        webpack: {
            aggregateTimeout: 300,
            poll: 1000
        }
    }
}
```

Sometimes it works well, but sometimes it just loop refresh.
 Maybe it's a webpack related issue?🤔

nuxt 2.12.2
webpack 4.42.0
node 14.3.0
npm 6.14.5

**Comment by yoyo837** [comment]
> Same issue
> 
> Only `nuxt.config.js` but issue still happened in Chrome and Firefox
> 
> ```js
> module.exports = {
>     srcDir: __dirname,
>     watchers: {
>         webpack: {
>             aggregateTimeout: 300,
>             poll: 1000
>         }
>     }
> }
> ```
> 
> Sometimes it works well, but sometimes it just loop refresh.
> Maybe it's a webpack related issue?🤔
> 
> nuxt 2.12.2
> webpack 4.42.0
> node 14.3.0
> npm 6.14.5

What 'Only Nuxt.config.js' means?

**Comment by DengSihan** [comment]
> > Same issue
> > Only `nuxt.config.js` but issue still happened in Chrome and Firefox
> > ```js
> > module.exports = {
> >     srcDir: __dirname,
> >     watchers: {
> >         webpack: {
> >             aggregateTimeout: 300,
> >             poll: 1000
> >         }
> >     }
> > }
> > ```
> > 
> > 
> > Sometimes it works well, but sometimes it just loop refresh.
> > Maybe it's a webpack related issue?🤔
> > nuxt 2.12.2
> > webpack 4.42.0
> > node 14.3.0
> > npm 6.14.5
> 
> What does 'Only Nuxt.config.js' mean?
My project only has a `nuxt.config.js`.
Sorry for poor English, I not a native specker

**Comment by simplenotezy** [comment]
> @valera-rozuvan As @manniL 's comment above [#3828 (comment)](https://github.com/nuxt/nuxt.js/issues/3828#issuecomment-464675623), it's actually a safari bug, as a workaround you may try to use @pi0 's solution.
> 
> Or change filename in dev mode (this may lead to memory leak as discussed above), like:
> 
> ```js
> // nuxt.config.js
> export default {
>   build: {
>     filenames: {
>       app: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js',
>       chunk: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js'
>     }
>   }
> }
> ```
> 
> As @pi0 is working on new architecture of server, so the memory leak may be fixed in v3, then the [hash] can be added as default value of dev bundle files.

I don't see @pi0's solution anywhere here. Anyhow, I tried to add your code to nuxt.config.js, but it doesn't solve the issue:

		build: {
			filenames: {
			app: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js',
			chunk: ({ isDev }) => isDev ? '[name].[hash].js' : '[chunkhash].js'
			}
		},

**Comment by simplenotezy** [comment]
Going incogntio-mode however seems to work..

**Comment by Jemaz** [comment]
> Going incogntio-mode however seems to work..

I can confirm that that infinite loop when hot reloading does not work in Firefox, but if it is incognito mode HMR works.

I haven't experienced the same issue on Chrome although (as in HMR works fine)

**Comment by theMangledBadger** [comment]
Running on:
Chrome: 83.0.4103.97
npm: 6.4.1
node: 10.15.0
nuxt: 2.12.2
webpack: 4.43.0

I had been experiencing this issue when I attempted to split my nuxt.config.js into separate config files, I've only started learning nuxt this week and I was probably trying to be too clever with it. 

HMR wouldn't pick up latest changes and initially running `npm run dev` would cause the app to reload 3-4 times before failing altogether.

Example exported config:

`const manifest = {
    name: 'XYZ',
    short_name: 'XYZ',
   etc....
};
  export default manifest;
`

And then in my nuxt.config.js, importing as
`import manifest from '~/config/manifest.js`

`export default {
    ... // standard nuxt config ,
    firebase,
    axios,
    **manifest**, 
    etc
`
I had tried clearing the cache, incognito mode, clearing application storage and Safari & Opera.

Reverting back to include everything in the nuxt.config.js file seems to have resolved the above issues.

I'm not sure if it is linked, but it appears to have been resolved for me this way, if anyone else has tried the same approach.

**Comment by mrkswrnr** [comment]
What finally worked for me was the following in my nuxt.config.js:
```
{
  build: {
    devMiddleware: {
      headers: {
        'Cache-Control': 'no-store',
        Vary: '*'
      }
    }
  }
}
```

**Comment by Grawl** [comment]
Just had the same issue in Quasar Framework. Open project in dev mode in Safari, got infinite reload with and without Web Inspector. I tried also turn on and off cache button in Web Inspector, has no effect.

Before, I had infinite reloading in Chrome with closed Web Inspector.

Then I googled it and I'm here. There is no solution for my problem, so I googled further.

I found this: https://stackoverflow.com/a/41936320/766307

Then I changed [`devtool`](https://webpack.js.org/configuration/devtool/) property to `eval-cheap-source-map` and problem is gone. Vue Devtools is working.

**Comment by carbdias** [comment]
I have this same issue without using pwa-module, just Nuxt 3 and Vue 3 and Vite, in 2023. I cannot refresh the page without getting an infinite loop in all browsers in Mac, and I always have to close the tab, clean cache and do a crtl+C and relaunch server again to see new results on the web.

**Comment by danielroe** [comment]
@carbdias I would suggest opening a new discussion rather than commenting on an issue closed 3 years ago (where it may unfortunately get lost)

**Comment by mathix420** [comment]
> I've been having this issue for the last few days but with chrome while using a cloud ide & a cloudflare tunnel.
> 
> The fix for me was to turn on "Development Mode" in cloudflare under "Caching > Configuration > Development Mode" which disables all caching on that side of things.
> 
> I assume this bug has something to do with a conflicting cache/header expectation for the dev server.

In my case this was not an option as my tunnel was defined in a sub domain used by other apps.
I've identified that the issue was also coming from `pwa-module`, so here is what I did to fix the issue:

1. Disable cache for the subdomain only using a cache rule: `Caching > Cache Rules > Create Rule > (http.host eq "my.sub.domain") > Bypass cache`
2. Redirect PWA manifest to a 404 page: `Rules > Transform Rules > Rewrite URL > Create Rule`
   - When incoming requests match `(http.host eq "my.sub.domain" and starts_with(http.request.uri.path, "/_nuxt/manifest."))`
   - Rewrite to > Static `_nuxt/404.json`
